### PR TITLE
drivers: sensor: Add NXP MCUX QDEC

### LINKED ISSUE #52211: MCUX based QDEC driver
**Is your enhancement proposal related to a problem? Please describe.**
I would like to read a quadrature encoded signal on an imxrt1052.

**Describe the solution you'd like**
Implement driver supporting NXP's QDEC driver.

**Additional context**
Link to NXP's driver:
https://github.com/zephyrproject-rtos/hal_nxp/tree/master/mcux/mcux-sdk/drivers/enc
Link to similar PR:
https://github.com/zephyrproject-rtos/zephyr/pull/49694

### PR DESCRIPTION
**Pull request by jeppenodgaard**
This PR adds support for Quadrature Encoder on the NXP MCUX.

Resolves https://github.com/zephyrproject-rtos/zephyr/issues/52211

### CODE REVIEW

**Comment by dleach02** [comment]
@jeppenodgaard, you will want to reference the west.yml update needed to pull in the HAL PR you submitted

**Comment by jeppenodgaard** [comment]
Ready for review.

Cannot be merge before https://github.com/zephyrproject-rtos/hal_nxp/pull/214 is merged and the nxp submodule is updated.

**Comment by danieldegrasse** [comment]
@jeppenodgaard, thank you for your contribution. Since this PR requires an update to the NXP HAL, the `west.yml` should be updated as part of this PR. The process of updating a west module is the following:

1. PR is submitted to module (in this case: https://github.com/zephyrproject-rtos/hal_nxp/pull/214)
2. PR is submitting to Zephyr with a `west.yml` referencing the relevant PR. For an example, see this west.yml file: https://github.com/nxp-zephyr/zephyr/blob/df7caa8b06b31c04a6b5c8c3c3d180e9e66b9680/west.yml#L91
3. PR is reviewed by Zephyr contributors
4. At some point, the PR to the module will be merged. For updates to NXP's HAL, we prefer to wait until the PR on the Zephyr side is ready to merge (or close to it) before merging the update to the HAL.
5. Once the module is updated, the `west.yml` file can be updated to point to the new SHA of the module's HAL. At this point the Zephyr PR can be updated, re-reviewed, and merged.

For more details on this process, see here: https://docs.zephyrproject.org/latest/develop/modules.html#process-for-submitting-changes-to-existing-modules

**Comment by jeppenodgaard** [comment]
@danieldegrasse thank you for helping out!
I changed the revision to the PR in the west.yml. So should be ready for step 3 now :crossed_fingers:

**Comment by danieldegrasse** [comment]
> I changed the revision to the PR in the west.yml. So should be ready for step 3 now 🤞

@jeppenodgaard- I think you may need to rebase onto a newer revision of `main`, it looks like the PR still has merge conflicts

**Comment by jeppenodgaard** [comment]
@danieldegrasse Should be ready now.

**Review by danieldegrasse — COMMENTED** [review]
You should split this PR into multiple commits. In the current form, this single commit may be difficult to bisect in the future.

Some guidelines for commits
- each commit must build cleanly
- commit should be split into logical changes

This is just a suggestion, but I might split the changes present here into commits doing the following: 
- Add QDEC driver into tree
- Add QDEC entries to RT10xx devicetree, and make any SOC level Kconfig changes that are required
- Add QDEC pinmux entries to RT1050 board
- Add QDEC sample

File: samples/sensor/qdec/src/main.c
Comment (jeppenodgaard) [reply samples/sensor/qdec/src/main.c]: Thanks! Fixed.

File: modules/Kconfig.mcux
Comment (jeppenodgaard) [reply modules/Kconfig.mcux]: Seems like it is unnecessary. Removed.

**Comment by jeppenodgaard** [comment]
@danieldegrasse PR has been split into logical commits and I have attempted to fix the issues you addressed.

File: dts/arm/nxp/nxp_rt10xx.dtsi
Comment (danieldegrasse) [inline dts/arm/nxp/nxp_rt10xx.dtsi:994]: I'd change the title of this commit to "dts: arm: add qdec nodes and update soc". The commit message also needs to be updated, as the `qdec` Kconfig is no longer added to the RT1050

File: samples/sensor/qdec/boards/mimxrt1050_evk.overlay
Comment (danieldegrasse) [inline samples/sensor/qdec/boards/mimxrt1050_evk.overlay:1]: I would refactor this commit to only include the addition of emulation support in the `qdec` sensor sample. Then, take the commit adding pin control support for `qdec `to the RT1050 EVK, and make that your final commit. You can include the overlay in that commit.

So to summarize, a series of commits like the following:
- add QDEC driver
- add DTS and Kconfig changes for RT1050
- add QDEC emulation support to sample
- add RT1050 pinmux and overlay to support QDEC sample

I'm aware this recommendation doesn't align with my prior request, I did not realize the QDEC sample already existed in tree on my first pass of this PR. If the QDEC sample is already upstream, I'd like to keep changes to the sample source code in a separate commit.

File: dts/arm/nxp/nxp_rt10xx.dtsi
Comment (jeppenodgaard) [reply dts/arm/nxp/nxp_rt10xx.dtsi:994]: Title and description updated

File: samples/sensor/qdec/boards/mimxrt1050_evk.overlay
Comment (jeppenodgaard) [reply samples/sensor/qdec/boards/mimxrt1050_evk.overlay:1]: @danieldegrasse No problem. I refactored the commits :+1:

**Review by danieldegrasse — DISMISSED** [review]

File: drivers/sensor/qdec_mcux/qdec_mcux.c
Comment (mmahadevan108) [inline drivers/sensor/qdec_mcux/qdec_mcux.c:142]: Can `XBARA1` be read from Devicetree as this can change to `XBARA` on certain platforms.

Comment (jeppenodgaard) [reply drivers/sensor/qdec_mcux/qdec_mcux.c]: I added it to the overlay. I would like to derive it e.g. from a xbar function but was unable to.

**Comment by jeppenodgaard** [comment]
@mmahadevan108 I think some tests are failing because https://github.com/zephyrproject-rtos/zephyr/pull/52730 has not been merged yet and the related PR in NXP HAL was merged after. Is that correct and is there a way to work around it?

**Comment by danieldegrasse** [comment]
> @mmahadevan108 I think some tests are failing because https://github.com/zephyrproject-rtos/zephyr/pull/52730 has not been merged yet and the related PR in NXP HAL was merged after. Is that correct and is there a way to work around it?

That does look to be correct- the RT600 uses variables declared in the `device_system` file (which implements CMSIS `SystemInit`). In the linked PR, the RT6xx enables, `CONFIG_PLATFORM_SPECIFIC_INIT`, which will include the `device_system` cmake file and resolve the build issue.


As far as resolutions that could allow this PR to move forwards without #52730 merging, another update would need to be applied to the NXP HAL that includes the `device_system` cmake file unconditionally if targeting the RT600. This PR would then pull in the NXP HAL at that revision

**Review by mmahadevan108 — DISMISSED** [review]

**Review by danieldegrasse — DISMISSED** [review]

**Comment by danieldegrasse** [comment]
> As far as resolutions that could allow this PR to move forwards without https://github.com/zephyrproject-rtos/zephyr/pull/52730 merging, another update would need to be applied to the NXP HAL that includes the device_system cmake file unconditionally if targeting the RT600. This PR would then pull in the NXP HAL at that revision

@jeppenodgaard I've opened a PR to make this fix in NXP's HAL: https://github.com/zephyrproject-rtos/hal_nxp/pull/218. Once this merges, you should be able to grab the latest SHA of the NXP's HAL and fix the CI failure.

File: drivers/sensor/qdec_mcux/qdec_mcux.c
Comment (MaureenHelm) [inline drivers/sensor/qdec_mcux/qdec_mcux.c:198]: Why not have a common init function shared among all instances?

File: dts/bindings/sensor/nxp,mcux-qdec.yaml
Comment (MaureenHelm) [inline dts/bindings/sensor/nxp,mcux-qdec.yaml:8]:
```suggestion
include: [pinctrl-device.yaml, sensor-device.yaml]
```

File: dts/bindings/sensor/nxp,mcux-qdec.yaml
Comment (MaureenHelm) [inline dts/bindings/sensor/nxp,mcux-qdec.yaml:29]: Should use phandles for this

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/sensor/qdec_mcux/qdec_mcux.c
Comment (mmahadevan108) [reply drivers/sensor/qdec_mcux/qdec_mcux.c]: @jeppenodgaard the `xbar` property has been added as part of `nxp,mcux-qdec.yaml` binding. Can this separated out to a separate node with its own binding inside `dts/bindings/arm`?

File: drivers/sensor/qdec_mcux/qdec_mcux.c
Comment (jeppenodgaard) [reply drivers/sensor/qdec_mcux/qdec_mcux.c:198]: There is a **counts_per_revolution** value for each qdec instance. I want to store the value directly in **qdec_config.revolutionCountCondition** to avoid having a duplicate value. But I cannot set the value before `ENC_GetDefaultConfig` is called.

File: drivers/sensor/qdec_mcux/qdec_mcux.c
Comment (jeppenodgaard) [reply drivers/sensor/qdec_mcux/qdec_mcux.c]: I think it would be nice to have a xbar driver. Then maybe inside the qdec node there would be a xbar child node that would be setup by the xbar driver so the qdec driver is more clean. Maybe this could be a seperate issue/PR?

Anyway if you want me to add it to this PR:
Do you want me to create e.g. `dts/bindings/arm/nxp,mcux-xbar.yaml` with a xbar property? I should also probably move xbar-maps?

File: drivers/sensor/qdec_mcux/qdec_mcux.c
Comment (MaureenHelm) [reply drivers/sensor/qdec_mcux/qdec_mcux.c:198]: You're duplicating an entire function to avoid duplicating counts_per_revolution (a uint32). You can store counts_per_revolution in the config struct similar to the base address.

File: drivers/sensor/qdec_mcux/qdec_mcux.c
Comment (mmahadevan108) [reply drivers/sensor/qdec_mcux/qdec_mcux.c]: I agree that having a driver would be a separate PR. I was thinking if we could add the device tree bindings as you have proposed above.

File: drivers/sensor/qdec_mcux/qdec_mcux.c
Comment (jeppenodgaard) [reply drivers/sensor/qdec_mcux/qdec_mcux.c:198]: I added the `count_per_revolution` to the config and tested the size.

Move to config with 1 instance:
```
           FLASH:       29148 B        64 MB      0.04%
             RAM:        5440 B        32 MB      0.02%
```
Duplicate init with 1 instance:
```
           FLASH:       29144 B        64 MB      0.04%
             RAM:        5440 B        32 MB      0.02%
```
Move to config with 2 instances:
```
           FLASH:       29308 B        64 MB      0.04%
             RAM:        5504 B        32 MB      0.02%
```
Duplicate init with 2 instances:
```
           FLASH:       29304 B        64 MB      0.04%
             RAM:        5504 B        32 MB      0.02%
```
It is nearly the same size.

Am I missing something?

Comment (jeppenodgaard) [reply drivers/sensor/qdec_mcux/qdec_mcux.c:198]: Build with debugging (CONFIG_DEBUG_OPTIMIZATIONS=y) is more efficient with the config approach:

Move to config with 2 instances and debug:
```
           FLASH:       32320 B        64 MB      0.05%
             RAM:        5504 B        32 MB      0.02%
```
Duplicate init with 2 instances and debug:
```
           FLASH:       32352 B        64 MB      0.05%
             RAM:        5504 B        32 MB      0.02%
```

File: drivers/sensor/qdec_mcux/qdec_mcux.c
Comment (jeppenodgaard) [reply drivers/sensor/qdec_mcux/qdec_mcux.c]: @mmahadevan108 I have tried to add your suggestion.

File: dts/bindings/sensor/nxp,mcux-qdec.yaml
Comment (jeppenodgaard) [reply dts/bindings/sensor/nxp,mcux-qdec.yaml]: @MaureenHelm fixed and reworked.

File: west.yml
Comment (jeppenodgaard) [reply west.yml]: @danieldegrasse I get [samples/synchronization failure](https://github.com/zephyrproject-rtos/zephyr/pull/53068/checks?check_run_id=10552146478) on boards `mimx8mp_evk_a53_smp` and `mimx93_evk_a55`. I am not sure but is there still a submodule issue?

**Comment by mmahadevan108** [comment]
@jeppenodgaard can you help address the CI flagged failures.

**Comment by jeppenodgaard** [comment]
> @jeppenodgaard can you help address the CI flagged failures.

I think it comes from some of the nxp module updates. I will try to revert the non-qdec commits and see if it fixes the CI failures.

**Comment by mmahadevan108** [comment]
We will help resolve the CI test related failures.

**Comment by mmahadevan108** [comment]
Test failures should be fixed by https://github.com/zephyrproject-rtos/hal_nxp/pull/222

**Comment by jeppenodgaard** [comment]
> Test failures should be fixed by [zephyrproject-rtos/hal_nxp#222](https://github.com/zephyrproject-rtos/hal_nxp/pull/222)

@mmahadevan108  Great. Do you want me to test the commit or wait until it is merged?

**Comment by jeppenodgaard** [comment]
> Test failures should be fixed by [zephyrproject-rtos/hal_nxp#222](https://github.com/zephyrproject-rtos/hal_nxp/pull/222)

@mmahadevan108  Updated hal_nxp to newest revision.

File: drivers/sensor/qdec_mcux/qdec_mcux.c
Comment (jeppenodgaard) [inline drivers/sensor/qdec_mcux/qdec_mcux.c:29]: @mmahadevan108  I fixed the other compliance errors but I cannot figure out why this has inconsistent spacing..?

**Review by MaureenHelm — DISMISSED** [review]

File: drivers/sensor/qdec_mcux/qdec_mcux.c
Comment (jeppenodgaard) [reply drivers/sensor/qdec_mcux/qdec_mcux.c:29]: @mmahadevan108 Do you want me to fiddle with it until the script does not fail or is it possible to merge because it is a false positive?

**Review by mmahadevan108 — DISMISSED** [review]

File: drivers/sensor/qdec_mcux/qdec_mcux.c
Comment (mmahadevan108) [reply drivers/sensor/qdec_mcux/qdec_mcux.c:29]: @nashif @stephanosio, can you please help with this Compliance Failure. Seems like a false positive.

**Comment by MaureenHelm** [comment]
Please rebase and resolve the merge conflict

**Comment by jeppenodgaard** [comment]
> Please rebase and resolve the merge conflict

@MaureenHelm done. Compliance failure is a false positive. See https://github.com/zephyrproject-rtos/zephyr/pull/53068#discussion_r1083051702.

**Review by mmahadevan108 — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**
